src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs

```diff
@@ -207,7 +207,10 @@ private unsafe bool TryUnblockSocket(bool abortive)
                 // Unless the user requested a normal close using Socket.Shutdown.
                 if (type == (int)SocketType.Stream && !_hasShutdownSend)
                 {
-                    Interop.Sys.Disconnect(handle);
+                    if (Interop.Sys.Disconnect(handle) == Interop.Error.EOPNOTSUPP)
+                    {
+                        Interop.Sys.Shutdown(handle, SocketShutdown.Both);
+                    }
```

Comment:
Could this instead be:
```C#
if (type != (int)SocketType.Stream ||
    _hasShutdownSend ||
    Interop.Sys.Disconnect(handle) != Interop.Error.EOPNOTSUPP)
{
    Interop.Sys.Shutdown(handle, SocketShutdown.Both);
}
```
?

Reply:
right. Right now we have  somewhat distinctive path stream vs datagram. And abortive vs non abortive in some other cases. We can compress the code but it _may_ Loose some clarity.  Maybe some comments? 
Maybe @tmds can chime in?

Reply:
Whichever you prefer

Reply:
`SystemNative_Disconnect` already uses/falls back to `shutdown`. We can move this change there by also using `shutdown` as a fallback on systems that `HAVE_DISCONNECTX`.

Reply:
updated.

Independently, neither of the calls update ` _hasShutdownSend` so it seems like we can call it over and over again. Is there some benefit to it @tmds?

Reply:
The reason for calling it again is for dealing with a race that might happen if an operation has already reffed the handle before we called `Dispose` and then starts the operation using that handle after we've called `TryUnblockSocket`. The next call to `TryUnblockSocket` is responsible for unblocking that operation (if it was blocking one).

With  `_hasShutdownSend` we're tracking if the user has explicitly called shutdown. If he did, we want the peer to see a regular TCP close.